src/Interpreters/RewriteCountDistinctVisitor.cpp

```diff
@@ -0,0 +1,59 @@
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <Interpreters/RewriteCountDistinctVisitor.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTLiteral.h>
+#include <Common/typeid_cast.h>
+#include "Coordination/KeeperStorage.h"
+#include "Parsers/ASTExpressionList.h"
+#include "Parsers/ASTIdentifier.h"
+#include "Parsers/ASTSelectQuery.h"
+#include "Parsers/ASTSubquery.h"
+#include "Parsers/ASTTablesInSelectQuery.h"
+#include <Parsers/Lexer.h>
+#include <Parsers/parseQuery.h>
+#include <Parsers/ParserQuery.h>
+
+namespace DB
+{
+
+void RewriteCountDistinctFunctionMatcher::visit(ASTPtr & ast, Data & /*data*/)
+{
+    auto * selectq = ast->as<ASTSelectQuery>();
+    if (!selectq || !selectq->tables() || selectq->tables()->children.size() != 1)
+        return;
+    auto expr_list = selectq->select();
+    if (!expr_list || expr_list->children.size() != 1)
+        return;
+    auto * func = expr_list->children[0]->as<ASTFunction>();
+    if (!func || (Poco::toLower(func->name) != "countdistinct" && Poco::toLower(func->name) != "uniqexact"))
+        return;
+    auto arg = func->arguments->children;
+    if (arg.size() != 1)
+        return;
+    if (!arg[0]->as<ASTIdentifier>())
+        return;
+    auto * table_expr = selectq->tables()->as<ASTTablesInSelectQuery>()->children[0]->as<ASTTablesInSelectQueryElement>()->children[0]->as<ASTTableExpression>();
+    if (!table_expr || !table_expr->database_and_table_name)
+        return;
+
+    auto column_name = arg[0]->as<ASTIdentifier>()->name();
+    func->name = "count";
+    func->children.clear();
+    func->children.emplace_back(std::make_shared<ASTExpressionList>());
+    func->arguments = func->children[0];
+    func->parameters = nullptr;
+
+    auto table_name = table_expr->database_and_table_name->as<ASTTableIdentifier>()->name();
+    table_expr->children.clear();
+    table_expr->children.emplace_back(std::make_shared<ASTSubquery>());
+    table_expr->database_and_table_name = nullptr;
+    table_expr->table_function = nullptr;
+    table_expr->subquery = table_expr->children[0];
+
+    std::string sub_sql = "SELECT " + column_name + " FROM " + table_name +  " GROUP BY " + column_name + "  ";
```

Comment:
seems like we know everything needed to build the AST without parsing. IMO it would be better to go without it.